### PR TITLE
Added collect rules to metrics allowlist

### PR DESCRIPTION
**Pull request by dislbenn**
Related issue: https://github.com/stolostron/backlog/issues/19141.
Signed-off-by: dislbenn <lavontae.bennett@gmail.com>

### CODE REVIEW

**Comment by dislbenn** [comment]
@marcolan018 here is the PR for adding the collect rules to the metrics allowlist. Here's a question, should we split the collect rules into a separate config map, or do we intend on maintaining the rules within the default allowlist?

File: operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml [hunk 1]
```diff
@@ -300,3 +300,53 @@ data:
...
+    - name: SNOHighMemoryUsage
+      expr: cluster:memory_usage:ratio{cluster_type\="SNO\"} * 100 > 80
+      for: 2m
+      metrics_list:
+        - container_memory_cache
+        - container_memory_rss
+        - container_memory_swap
+        - container_memory_working_set_bytes
+        - kube_pod_container_resource_limits
+        - kube_pod_container_resource_requests
```
Comment (dislbenn) [inline operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml:352]: As recommended, we need two collect rules for SNO clusters also. We need to decide if the metrics being collected will also follow the same rule as the namespace cluster.

File: operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml [hunk 2]
```diff
@@ -300,3 +300,53 @@ data:
         expr: sum(rate(node_network_receive_drop_total{job=\"node-exporter\", device!=\"lo\"}[1m])) without(device)
       - record: instance:node_network_transmit_drop_excluding_lo:rate1m
         expr: sum(rate(node_network_transmit_drop_total{job=\"node-exporter\", device!=\"lo\"}[1m])) without(device)
+  metrics_rules.yaml: |
+    - name: HighCPUUsage
```
Comment (marcolan018) [inline operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml:304]: better change name to "NamespaceHighCPUUsage"

File: operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml [hunk 3]
```diff
@@ -300,3 +300,53 @@ data:
...
+      expr: cluster:container_cpu_usage:ratio * 100 > 80
```
Comment (marcolan018) [inline operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml:305]: should use metrics "kube_resourcequota" in the expr

File: operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml [hunk 4]
```diff
@@ -300,3 +300,53 @@ data:
...
+      metrics_list:
+        - container_memory_cache
+        - container_memory_rss
+        - container_memory_swap
+        - container_memory_working_set_bytes
+        - kube_pod_container_resource_limits
+        - kube_pod_container_resource_requests
+
+    - name: SNOHighCPUUsage
+      expr: cluster:container_cpu_usage:ratio{cluster_type\="SNO\"} * 100 > 80
```
Comment (marcolan018) [inline operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml:330]: cluster_type label is not available in managed cluster side and cannot be used in  collect rule. 
Need to think out a way for this
one option is to add a "selector" field like
```
selector:
  clusterType: "SNO"
```

File: operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml (hunk 1 above)
Comment (dislbenn) [reply operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml]: SNO collect rules will be prefixed with `SNO`; whereas, namespace cluster rules will contain `Namespace`.

File: operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml [hunk 5]
```diff
@@ -300,3 +300,57 @@ data:
...
+        mixin_pod_workload
+        namespace_workload_pod:kube_pod_owner:relabel
+        node_namespace_pod_container:container_cpu_usage_seconds_total:sum_irate
+        node_namespace_pod_container:container_cpu_usage_seconds_total:sum_rate
+
+    - name: NamespaceHighMemoryUsage
+      expr: cluster:memory_usage:ratio * 100 > 80
+      for: 2m
+      metrics_list:
+        - container_memory_cache
```
Comment (dislbenn) [inline operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml:322]: Reduce 80 to 70.

Comment (dislbenn) [reply operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml:322]: Replace expr with kube_resourcequota.

**Comment by dislbenn** [comment]
@marcolan018 I updated the expressions for the collect rules, can you review them?

File: operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml [hunk 6]
```diff
@@ -300,3 +300,57 @@ data:
         expr: sum(rate(node_network_receive_drop_total{job=\"node-exporter\", device!=\"lo\"}[1m])) without(device)
       - record: instance:node_network_transmit_drop_excluding_lo:rate1m
         expr: sum(rate(node_network_transmit_drop_total{job=\"node-exporter\", device!=\"lo\"}[1m])) without(device)
+  metrics_rules.yaml: |
+    - name: NamespaceHighCPUUsage
+      expr: (sum(kube_resourcequota{resource=\"limits.cpu\", type=\"used\"}) by (cluster, namespace, resourcequota) / sum(kube_resourcequota{resource=\"limits.cpu\", type=\"hard\"}) by (cluster, namespace, resourcequota)) * 100 > 70
```
Comment (marcolan018) [inline operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml:305]: cluster label is not required in expr

Comment (dislbenn) [reply operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml]: Removed the cluster label.

Comment (dislbenn) [reply operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml]: @marcolan018 are there any other fields that we would want to capture?

Comment (marcolan018) [reply operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml]: no, I think current expr is good

File: operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml [hunk 7]
```diff
@@ -300,3 +300,57 @@ data:
...
+      for: 2m
+      selector:
+        clusterType: "SNO"
+      metrics_list:
+        - container_memory_cache
+        - container_memory_rss
+        - container_memory_swap
+        - container_memory_working_set_bytes
+        - kube_pod_container_resource_limits
+        - kube_pod_container_resource_requests
```
Comment (marcolan018) [inline operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml:356]: where is the metrics_list come from?
ideally, they should cover all the pod/container level metrics in dashboards below:
Kubernetes / Compute Resources / Namespace (Pods)
Kubernetes / Compute Resources / Namespace (Workloads)
Kubernetes / Compute Resources / Nodes (Pods)
Kubernetes / Compute Resources / Pod
Kubernetes / Compute Resources / Workload

Comment (dislbenn) [reply operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml:356]: The metrics that are listed are from those dashboards. Here is the comment that captures the metrics: https://github.com/stolostron/backlog/issues/19141#issuecomment-1050186409.

**Comment by dislbenn** [comment]
@marcolan018 I've just updated the metrics list for the collect rules. This should cover the remaining changes that we needed to merge this story. Can you help review it?

File: operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml [hunk 8]
```diff
@@ -300,3 +300,63 @@ data:
         expr: sum(rate(node_network_receive_drop_total{job=\"node-exporter\", device!=\"lo\"}[1m])) without(device)
       - record: instance:node_network_transmit_drop_excluding_lo:rate1m
         expr: sum(rate(node_network_transmit_drop_total{job=\"node-exporter\", device!=\"lo\"}[1m])) without(device)
+  metrics_rules.yaml: |
```
Comment (marcolan018) [inline operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml:303]: this line should be 
```
collect_rules:
```

File: operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml [hunk 9]
```diff
@@ -300,3 +300,63 @@ data:
...
+      metrics_list:
+        - container_memory_cache 
+        - container_memory_rss 
+        - container_memory_swap 
+        - container_memory_working_set_bytes 
+        - kube_pod_container_resource_limits 
+        - kube_pod_container_resource_requests 
+        - kube_resourcequota 
+        - mixin_pod_workload 
+        - namespace_workload_pod:kube_pod_owner:relabel
```
Comment (marcolan018) [inline operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml:330]: For namespace collect rule, seems we need more enhancements.
e.g. in a large cluster, if we detect some problem in namespace A, should only start to collect the metrics in namespace A, not other namespaces.
If that, the metrics_list should be changed to something like below:
```
    matches:
      - __name__="container_memory_cache",namespace="{{ $labels.namespace }}"
```
by the way, to be consistent with allowlist, the metrics_list in SNO rules should change to "names"

Comment (dislbenn) [reply operators/multiclusterobservability/manifests/base/config/metrics_allowlist.yaml]: @marcolan018 I updated the metrics list, can you confirm if the changes are correct?

**Pull request merged**